### PR TITLE
add tracking of pack environments for pack elements to Constraint System

### LINKED ISSUE #67503: Using value pack in closure crashes compiler
This code crashes the compiler:

```swift
struct SyncedTypes<T1, T2> {
	let properties: [(T1, inout T2) -> Void]
	
	init<each V: Equatable>(_ pair: repeat (KeyPath<T1, each V>, WritableKeyPath<T2, each V>)) {
		var properties = [(T1, inout T2) -> Void]()

		repeat properties.append( { (source: T1, destination: inout T2) in
			destination[keyPath: each pair.1] = source[keyPath: each pair.0]
		})

		self.properties = properties
	}
}
```

**Environment**

Xcode: Version 15.0 beta 4 (15A5195m)

Deployment target: iOS 17.0

### PR DESCRIPTION
**Pull request by sophiapoirier**
This refactor is a prerequisite to my next PR that will address an abstract tuple bug.

### CODE REVIEW

**Comment by sophiapoirier** [comment]
@swift-ci please test

**Comment by sophiapoirier** [comment]
@swift-ci please test

File: lib/Sema/CSGen.cpp [hunk 1]
```diff
@@ -3061,9 +3084,10 @@ namespace {
 
       for (auto pack : expandedPacks) {
         Type packType;
-        if (auto *elementExpr = getAsExpr<PackElementExpr>(pack)) {
-          packType = CS.getType(elementExpr->getPackRefExpr());
-        } else if (auto *elementType = getAsTypeRepr<PackElementTypeRepr>(pack)) {
+        if (isExpr<PackElementExpr>(pack)) {
+          continue;
```
Comment (xedin) [inline lib/Sema/CSGen.cpp:3088]: Nit: Could you please leave a comment here explaining why we skip over pack elements and where the connection between this pack expansion and pack element is established now to make it easier in the future to figure out what is going on here?

**Review by xedin — APPROVED** [review]
LGTM! The comment I left shouldn't block you from merging.

**Comment by sophiapoirier** [comment]
@swift-ci please test

**Comment by sophiapoirier** [comment]
@swift-ci please test

**Pull request merged**